File: fixtures/ts-sample/src/async_calls.ts
// cgx-fixture: async/await (TypeScript)
// Covers: calls:async (await on async fn), suspends property at await sites,
//         Promise chains as calls:async, async method on class

export async function fetchData(url: string): Promise<string> {
    const raw = await httpGet(url);        // calls:async, suspension point
    const parsed = await parseResponse(raw);  // calls:async, suspension point
    return parsed;
}

async function httpGet(url: string): Promise<string> {
    if (!url) throw new Error("empty url");
    return `response from ${url}`;
}

async function parseResponse(raw: string): Promise<string> {
    if (!raw) throw new Error("empty response");
    return raw.toUpperCase();
}

export class ApiClient {
    private baseUrl: string;

    constructor(baseUrl: string) {
        this.baseUrl = baseUrl;
    }

    async get(path: string): Promise<string> {
        const url = `${this.baseUrl}/${path}`;
        return fetchData(url);  // calls:async (implicit await in async method)
    }

    async post(path: string, body: string): Promise<string> {
        const url = `${this.baseUrl}/${path}`;
        return httpPost(url, body);   // calls:async
    }
}

async function httpPost(url: string, body: string): Promise<string> {
    if (!url || !body) throw new Error("invalid request");
    return `posted ${body.length} bytes to ${url}`;
}

/// Sequential awaits — each is a separate suspension point.
export async function sequentialAwaits(url: string): Promise<string> {
    const a = await httpGet(url);   // suspension point 1
    const b = await httpGet(url);   // suspension point 2
    return a + b;
}

/// await in a try block — suspension point inside exception context.
export async function awaitInTry(url: string): Promise<string> {
    try {
        const result = await httpGet(url);   // calls:async, always (inside try body)
        return result;
    } catch (e) {
        return "error";
    }
}

/// Promise.all — awaits multiple promises — calls:async on the await.
export async function parallel(urls: string[]): Promise<string[]> {
    const promises = urls.map(url => httpGet(url));
    return Promise.all(promises);  // calls:async on the await of Promise.all
}

File: fixtures/rust-sample/src/async_calls.rs
// cgx-fixture: async/await
// Covers: calls:async (await on async fn), suspends property at await sites,
//         async method on struct

use std::time::Duration;

pub async fn fetch_data(url: &str) -> Result<String, String> {
    let raw = http_get(url).await?;
    let parsed = parse_response(&raw).await?;
    Ok(parsed)
}

async fn http_get(url: &str) -> Result<String, String> {
    // Simulated HTTP call — in real code this would be reqwest or hyper
    if url.is_empty() {
        return Err("empty url".to_string());
    }
    Ok(format!("response from {}", url))
}

async fn parse_response(raw: &str) -> Result<String, String> {
    if raw.is_empty() {
        Err("empty response".to_string())
    } else {
        Ok(raw.to_uppercase())
    }
}

pub struct AsyncService {
    base_url: String,
}

impl AsyncService {
    pub fn new(base_url: &str) -> Self {
        AsyncService { base_url: base_url.to_string() }
    }

    pub async fn get(&self, path: &str) -> Result<String, String> {
        let url = format!("{}/{}", self.base_url, path);
        fetch_data(&url).await
    }

    pub async fn post(&self, path: &str, body: &str) -> Result<String, String> {
        let url = format!("{}/{}", self.base_url, path);
        http_post(&url, body).await
    }
}

async fn http_post(url: &str, body: &str) -> Result<String, String> {
    if url.is_empty() || body.is_empty() {
        Err("invalid request".to_string())
    } else {
        Ok(format!("posted {} bytes to {}", body.len(), url))
    }
}

/// Awaiting multiple futures sequentially — each await is a suspension point.
pub async fn sequential_awaits(url: &str) -> String {
    let a = http_get(url).await.unwrap_or_default();
    let b = http_get(url).await.unwrap_or_default();
    format!("{}{}", a, b)
}

/// sleep-like simulation — suspension without a meaningful return value.
pub async fn delay_then_run(millis: u64) -> String {
    simulate_sleep(Duration::from_millis(millis)).await;
    "done".to_string()
}

async fn simulate_sleep(_d: Duration) {
    // placeholder — would be tokio::time::sleep in real code
}
